lib/core/src/Cardano/Pool/Metadata.hs

```diff
@@ -65,23 +95,123 @@ type Api
 -- | A client for fetching metadata from an Aggregation server.
 --
 -- See also 'newClient' to construct a client.
-newtype Client api = Client
+newtype Client api m = Client
     { getStakePoolMetadata
         :: PoolId
-        -> IO (Either ClientError StakePoolOffChainMetadata)
+        -> m (Either ClientError (Maybe StakePoolOffChainMetadata))
     }
 
-mkClient
+mkClientIO
     :: Manager
     -> BaseUrl
-    -> Client Api
-mkClient mgr baseUrl = Client
-    { getStakePoolMetadata = \pid ->
-        fmap getApiT <$> run (getMetadata (ApiT pid))
+    -> Client Api IO
+mkClientIO mgr baseUrl = Client
+    { getStakePoolMetadata = \pid -> do
+        run (getMetadata (ApiT pid)) <&> \case
+            Right (ApiT meta) ->
+                Right (Just meta)
+
+            Left (FailureResponse _ res) | responseStatusCode res == status404 -> do
+                Right Nothing
+
+            Left e ->
+                Left e
     }
   where
     run :: ClientM a -> IO (Either ClientError a)
     run query = runClientM query (mkClientEnv mgr baseUrl)
 
     getMetadata =
         client (Proxy @Api)
+
+-- | A configuration for managing metadata with the aggregation server.
+-- Callbacks and parameterized effects allows for easier testing while a real
+-- specialization would wire a database connector in here.
+data MetadataConfig (m :: * -> *) = MetadataConfig
+    { saveMetadata
+        :: PoolId -> (Maybe StakePoolOffChainMetadata, UTCTime) -> m ()
+        -- ^ A callback action for storing an off-chain metadata. The callback
+        -- may be called with 'Nothing' to store that no metadata were found for
+        -- a particular 'PoolId; this allows for not constantly re-fetching data
+        -- for pools that are known to have no metadata.
+
+    , getModificationTime
+        :: PoolId -> m (Maybe UTCTime)
+        -- ^ Action for fetching the last modification time of a cached result.
+        -- 'Nothing' is expected when there's no cached result.
+
+    , apiClient
+        :: Client Api m
+        -- ^ A client constructed with 'mkClient' for interacting with a
+        -- metadata aggregation server.
+
+    , cacheTTL
+        :: NominalDiffTime
+        -- ^ A constant for the maximum age of cached registry metadatabefore
+        -- it's considered to be stale.
+    }
+
+-- | Refresh metadata for a list of pool ids. Where metadata are cached is
+-- out of the scope for this function and provided through callbacks.
+--
+-- TODO: Add some typed log messages in here.
+refresh
+    :: Tracer IO MetadataRegistryLog
+        -- ^ Logging object for capturing events in the 'refresh' function.
+        --
+    -> MetadataConfig IO
+        -- ^ Configuration and callbacks for persistence of metadata.
+
+    -> [PoolId]
+        -- ^ A list of pool ids for which metadata need to be fetched or
+        -- refreshed.
+    -> IO ()
```

Comment:
Why is `refresh` needed? Can we not just put the caching inside `Client`?
```haskell
cacheClient 
    :: (PoolId -> m (Maybe StakePoolOffChainMetadata, UTCTime)) 
    -> (PoolId -> Maybe (StakePoolOffChainMetadata, UTCTime) -> m ()) 
    -> Client Api m 
    -> Client Api m
cacheClient pid putCached getCached rawClient = Client { getStakePoolMetadata
    if stale 
    then (getStakePoolMetadata rawClient) >>= \x -> putCache x >> return x 
    else getCached
}
```

Reply:
That is a valid point. Felt simpler to write both decoupled initially but, simply having a cached client make sense.

Reply:
I was more curious about a `Client Api m -> Client Api m` function, but sure, what you have now works too.